model/types.py

```diff
@@ -52,42 +55,149 @@ class MentoExchange(SerializableEnum):
 
 
 class Pair(NamedTuple):
+    """
+    Base Class for Exchange Pairs
+    """
     base: Currency
     quote: Currency
 
+    # pylint: disable = missing-function-docstring;
     def __str__(self):
         return f"{self.base.value}_{self.quote.value}"
+    # pylint: disable = unsubscriptable-object;
+
+    def get_rate(self, prev_state) -> float:
+        market_prices = prev_state['market_price']
+
+        if self in market_prices:
+            rate = market_prices.get(self)
+
+        elif isinstance(self.base, Fiat) and isinstance(self.quote, Fiat):
+            rate = self.fiat_to_fiat(prev_state)
+
+        elif isinstance(self.base, CryptoAsset) and isinstance(self.quote, CryptoAsset):
+
+            rate = self.crypto_to_crypto(prev_state)
+
+        elif isinstance(self.base, Stable) and isinstance(self.quote, Stable):
+
+            rate = self.stable_to_stable(prev_state)
+
+        elif isinstance(self.base, CryptoAsset) and isinstance(self.quote, Fiat):
+            rate = self.crypto_to_fiat(prev_state)
+
+        elif isinstance(self.base, CryptoAsset) and isinstance(self.quote, Stable):
+
+            rate = self.crypto_to_stable(prev_state)
+
+        elif isinstance(self.base, Stable) and isinstance(self.quote, Fiat):
+
+            rate = self.stable_to_fiat(prev_state)
+
+        else:
+            pair_reversed = Pair(self.quote, self.base)
+            rate = pair_reversed.get_rate(prev_state)**(-1)
+
+        return rate
```

Comment:
I have an idea to simplify things here but it's optional if you want it (this is a sketch): 
```python

class Rate(NamedTuple):
  value: float
  pair: Pair

  def __mul__(self, other: Rate):
    assert self.pair.quote == other.pair.base, "Pairs don't match"
    return Rate(self.value * other.value, Pair(self.pair.base, other.pair.quote))
   
class Pair(NamedTuple):
    # ...
    @property
    def inverse(self) -> Pair:
        return Pair(self.quote, self.base)

    def get_rate(self, state: StateVariables) -> Rate:
        if self in state['market_prices']:
            return state['market_prices'].get(self)
        elif self.inverse in state['market_prices']:
            return 1 / state['market_prices'].get(self.inverse) 
        
        bridge_currency = self.get_bridge_currency()
        base_to_bridge = Pair(self.base, bridge_currency)
        bridge_to_quote = Pair(bridge_currency, self.quote)
        return base_to_bridge.get_rate(state) * bridge_to_quote.get_rate(state)
    
    def get_bridge_currency(self) -> Currency:
        if isinstance(self.base, Fiat) and isinstance(self.quote, Fiat):
            return CryptoAsset.CELO
        elif isinstance(self.base, CryptoAsset) and isinstance(self.quote, CryptoAsset):
            return Fiat.USD
        elif isinstance(self.base, Stable) and isinstance(self.quote, Stable):
            # Didn't fully understand your strategy here
            return something
        elif isinstance(self.base, CryptoAsset) and isinstance(self.quote, Fiat):
            return CryptoAsset.CELO
        elif isinstance(self.base, CryptoAsset) and isinstance(self.quote, Stable):
            # etc...
```

There are two things I really like with this approach:
1. By introducing the idea of the inverse it pretty much removes the need for that 4-way if you do in `implicit_rate`
2. It makes the logic easier to follow, because what you actually need to do when determining a rate is one of these three things:
  (a) Return the rate if the pair exists in `market_prices`
  (b) Return 1 / rate if the inverse of the pair exists in market_prices
  (c) Find a bridge asset for which `(self.base, bridge)` (or it's inverse) and `(bridge, self.quote)` (or it's inverse) exist in `market_prices`

And that's essentially what you're doing as well.

Reply:
```python
elif isinstance(self.base, Stable) and isinstance(self.quote, Stable):
            # Didn't fully understand your strategy here
            return something
```
This case need a `bridge implicit pair` cUSD/USD -> USD/CELO -> CELO/EUR -> EUR/cEUR